### PR TITLE
Determine the position of a Coordinate relative to a geometry.

### PR DESCRIPTION
**Pull request by michaelkirk**
# Examples
```rust
use geo::{polygon, Coordinate};
use geo::coordinate_position::{CoordinatePosition, CoordPos};
let square_poly = polygon![(x: 0.0, y: 0.0), (x: 2.0, y: 0.0), (x: 2.0, y: 2.0), (x: 0.0, y: 2.0), (x: 0.0, y: 0.0)];
let inside_coord = Coordinate { x: 1.0, y: 1.0 };
assert_eq!(square_poly.coordinate_position(&inside_coord), CoordPos::Inside);
let boundary_coord = Coordinate { x: 0.0, y: 1.0 };
assert_eq!(square_poly.coordinate_position(&boundary_coord), CoordPos::OnBoundary);
let outside_coord = Coordinate { x: 5.0, y: 5.0 };
assert_eq!(square_poly.coordinate_position(&outside_coord), CoordPos::Outside);
```

Some work extracted from the geomgraph effort.

### CODE REVIEW

File: geo-types/src/coordinate.rs [hunk 1]
```diff
@@ -222,14 +222,19 @@ use num_traits::Zero;
 /// assert_eq!(p.x, 0.);
 /// assert_eq!(p.y, 0.);
 /// ```
-impl<T: CoordinateType> Zero for Coordinate<T> {
-    fn zero() -> Self {
+impl<T: CoordinateType> Coordinate<T> {
+    pub fn zero() -> Self {
```
Comment (michaelkirk) [inline geo-types/src/coordinate.rs:226]: ...tangential quality of life improvement.

File: geo/src/algorithm/coordinate_position.rs [hunk 2]
```diff
@@ -0,0 +1,749 @@
...
+            Geometry::GeometryCollection(geometry_collection) => {
+                geometry_collection.calculate_coordinate_position(coord, is_inside, boundary_count)
+            }
+        }
+    }
+}
+
+/// Calculate the position of a `Coordinate` relative to a
+/// closed `LineString`.
+pub fn coord_pos_relative_to_ring<T>(coord: Coordinate<T>, linestring: &LineString<T>) -> CoordPos
```
Comment (michaelkirk) [inline geo/src/algorithm/coordinate_position.rs:362]: This method was moved from utils unmodified.

In an effort to keep `utils` from becoming a big junk drawer, this seemed like a more specific home for it.

File: geo/src/algorithm/coordinate_position.rs [hunk 3]
```diff
@@ -0,0 +1,749 @@
...
+    kernels::HasKernel,
+};
+use crate::{
+    Coordinate, Geometry, GeometryCollection, Line, LineString, MultiLineString, MultiPoint,
+    MultiPolygon, Point, Polygon, Rect, Triangle,
+};
+
+/// The position of a `Coordinate` relative to a `Geometry`
+#[derive(PartialEq, Clone, Debug)]
+pub enum CoordPos {
```
Comment (michaelkirk) [inline geo/src/algorithm/coordinate_position.rs:12]: moved this from utils - reusing form the "coord_pos_relative_to_ring".

@rmanoka - do you see any issue with re-using this enum from coord_pos_relative_to_ring? 

At first I introduced a new "Interior/Exterior/Boundary" enum, but it seemed redundant.

Comment (rmanoka) [reply geo/src/algorithm/coordinate_position.rs:12]: Sounds fine to me

File: geo/src/algorithm/coordinate_position.rs [hunk 4]
```diff
@@ -0,0 +1,749 @@
...
+/// let inside_coord = Coordinate { x: 1.0, y: 1.0 };
+/// assert_eq!(square_poly.coordinate_position(&inside_coord), CoordPos::Inside);
+///
+/// let boundary_coord = Coordinate { x: 0.0, y: 1.0 };
+/// assert_eq!(square_poly.coordinate_position(&boundary_coord), CoordPos::OnBoundary);
+///
+/// let outside_coord = Coordinate { x: 5.0, y: 5.0 };
+/// assert_eq!(square_poly.coordinate_position(&outside_coord), CoordPos::Outside);
+/// ```
+pub trait CoordinatePosition<T>
```
Comment (rmanoka) [inline geo/src/algorithm/coordinate_position.rs:37]: @michaelkirk Do you prefer the `CoordinatePosition<T>` in favor of `CoordinatePosition` with a associate type `Scalar: CoordinateType` ?  My personal pref. is the latter;  optional comment.

File: geo/src/algorithm/coordinate_position.rs [hunk 5]
```diff
@@ -0,0 +1,749 @@
...
+    T: HasKernel,
+{
+    fn calculate_coordinate_position(
+        &self,
+        coord: &Coordinate<T>,
+        is_inside: &mut bool,
+        boundary_count: &mut usize,
+    ) {
+        // PERF TODO: I'm sure there's a better way to calculate than converting to a polygon
+        Polygon::from(self.clone()).calculate_coordinate_position(coord, is_inside, boundary_count);
```
Comment (frewsxcv) [inline geo/src/algorithm/coordinate_position.rs:184]: i think you can do:

```suggestion
        self.to_polygon().calculate_coordinate_position(coord, is_inside, boundary_count);
```

File: geo/src/algorithm/coordinate_position.rs [hunk 6]
```diff
@@ -0,0 +1,749 @@
...
```
Comment (frewsxcv) [inline geo/src/algorithm/coordinate_position.rs:199]:
```suggestion
        self.to_polygon().calculate_coordinate_position(coord, is_inside, boundary_count);
```

File: geo/src/algorithm/coordinate_position.rs [hunk 7]
```diff
@@ -0,0 +1,749 @@
...
+where
+    T: HasKernel,
+{
+    fn coordinate_position(&self, coord: &Coordinate<T>) -> CoordPos {
+        let mut is_inside = false;
+        let mut boundary_count = 0;
+
+        self.calculate_coordinate_position(coord, &mut is_inside, &mut boundary_count);
+
+        // SFS "mod 2" boundary rule
```
Comment (frewsxcv) [inline geo/src/algorithm/coordinate_position.rs:47]: thoughts on adding a quote here?

```suggestion
        // “The boundary of an arbitrary collection of geometries whose interiors are disjoint
        // consists of geometries drawn from the boundaries of the element geometries by
        // application of the ‘mod 2’ union rule”
        //
        // ― OpenGIS Simple Feature Access § 6.1.15.1
```

**Review by frewsxcv — APPROVED** [review]
looks great, thanks!

File: geo/src/algorithm/coordinate_position.rs (hunk 4 above)
Comment (michaelkirk) [reply geo/src/algorithm/coordinate_position.rs]: I believe I've done this here: a60dc27 - can you take a look @rmanoka?

Despite reading a couple articles on this, I only kind of theoretically understand why this distinction is good, and still don't have a very intuitive grasp for when to apply it.

Is your rule of thumb to prefer associated types and only use full-blown generic type params when associated types are insufficient?

Comment (michaelkirk) [reply geo/src/algorithm/coordinate_position.rs]: I'll plan to merge this in a couple days if I don't hear otherwise.

**Comment by michaelkirk** [comment]
bors r=frewsxcv

**Comment by michaelkirk** [comment]
bors retrt

**Comment by michaelkirk** [comment]
bors retry

File: geo/src/algorithm/coordinate_position.rs (hunk 4 above)
Comment (rmanoka) [reply geo/src/algorithm/coordinate_position.rs]: @michaelkirk Sorry for the late response;  the changes look great!   My pref. is mostly similar to `Iterator` vs `AddAsign` traits in std.  I go for an associated type unless the same type could implement the trait with two possible param.  Here, we're implementing the trait `CoordinatePosition` with only `T = Ty` for `Line<Ty>`, so prefer it to be associated than make the trait `CoordinatePosition<T>`.  

May be in the future, our scalar types could have additional sophistication that tells us when `T1` can be converted to `T2` without loss in precision, and we can make some of our traits like `CoordinatePosition`, `ForEach` into generics.

**Pull request merged**